utils/src/test/java/com/cloud/utils/testcase/NioTest.java

```diff
@@ -19,146 +19,198 @@
-        try {
-            _server.start();
-        } catch (final NioConnectionException e) {
-            fail(e.getMessage());
-        }
+        completedTestCount = 0;
+        testBytes = new byte[1000000];
+        randomGenerator.nextBytes(testBytes);
 
-        _client = new NioClient("NioTestServer", "127.0.0.1", 7777, 5, new NioTestClient());
+        // Server configured with one worker
+        server = new NioServer("NioTestServer", 0, 1, new NioTestServer());
         try {
-            _client.start();
+            server.start();
         } catch (final NioConnectionException e) {
-            fail(e.getMessage());
+            Assert.fail(e.getMessage());
         }
 
-        while (_clientLink == null) {
-            try {
-                s_logger.debug("Link is not up! Waiting ...");
-                Thread.sleep(1000);
-            } catch (final InterruptedException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
+        // 5 malicious clients per valid client
+        for (int i = 0; i < totalTestCount; i++) {
+            for (int j = 0; j < 5; j++) {
+                final NioClient maliciousClient = new NioMaliciousClient("NioMaliciousTestClient-" + i, "127.0.0.1", server.getPort(), 1, new NioMaliciousTestClient());
+                maliciousClients.add(maliciousClient);
+                maliciousExecutor.submit(new ThreadedNioClient(maliciousClient));
             }
+            final NioClient client = new NioClient("NioTestClient-" + i, "127.0.0.1", server.getPort(), 1, new NioTestClient());
+            clients.add(client);
+            clientExecutor.submit(new ThreadedNioClient(client));
         }
     }
 
-    @Override
+    @After
     public void tearDown() {
+        stopClient();
+        stopServer();
+    }
+
+    protected void stopClient() {
+        for (NioClient client : clients) {
+            client.stop();
+        }
+        for (NioClient maliciousClient : maliciousClients) {
+            maliciousClient.stop();
+        }
+        LOGGER.info("Clients stopped.");
+    }
+
+    protected void stopServer() {
+        server.stop();
+        LOGGER.info("Server stopped.");
+    }
+
+    @Test
+    public void testConnection() {
+        final long currentTime = System.currentTimeMillis();
         while (!isTestsDone()) {
+            if (System.currentTimeMillis() - currentTime > 600000) {
+                Assert.fail("Failed to complete test within 600s");
+            }
             try {
-                s_logger.debug(_completedCount + "/" + _testCount + " tests done. Waiting for completion");
+                LOGGER.debug(completedTestCount + "/" + totalTestCount + " tests done. Waiting for completion");
                 Thread.sleep(1000);
             } catch (final InterruptedException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
+                Assert.fail(e.getMessage());
```

Comment:
@bhaisaab as the *completedTestCount* is incremented at the *oneMoreTestDone()* method, I think that this log could be moved to "*com.cloud.utils.testcase.NioTest.oneMoreTestDone()*"; this would remove the need of the *Thread.sleep(1000)* and let the log clean.